improvement(Improve Employees Standings report)

lomamech opened the pull request:
- Add the possibility to visualize the employees standings by range of
  date, by the adding of component date by bh-date-interval
- Add filter by range of date

closes #3796

lomamech committed:
improvement(Improve Employees Standings report)
- Add the possibility to visualize the employees standings by range of
  date, by the adding of component date by bh-date-interval
- Add filter by range of date

closes #3796

lomamech committed:
refactor(Employee Standing)
- Refactor and Sanitaze code

lomamech committed:
Sanitaze code

jniles commented on client/src/i18n/en/form.json:
```diff
@@ -402,6 +402,7 @@
             "ERROR_404": "Error 404",
             "EXCHANGE_RATE": "Exchange Rate",
             "EXPENSE": "Expenses",
+            "EXTRANCT_EMPLOYEE_STANDINGS": "Extract from Employee Standing of period",
```
This should be "Extract of the Employee Standing by Period"

jniles commented on client/src/i18n/en/table.json:
```diff
@@ -103,6 +103,7 @@
             "EXCHANGE_RATE": "Exchange Rate",
             "EXPIRATION_DATE": "Expiration Date",
             "EXPIRE_IN":"Expire In",
+            "EXTRACT_EMPLOYEE_STANDINGS": "View extract of the employee's standings by date range",
```
This is almost perfect!  You just need to change "standings" -> "standing".  It isn't plural in English.

jniles commented on server/controllers/finance/reports/financial.employee.js:
```diff
@@ -71,8 +76,30 @@ async function build(req, res, next) {
+    // also be the employee's balance for the search period
+    if (options.extractEmployee) {
+      if (creditorOperations.transactions.length) {
+        data.lastCum = creditorOperations.transactions[creditorOperations.transactions.length - 1];
```
Let's call this `lastTransaction` or `lastTxn` or something like that since it has the entire transaction and not just the cumsum.

jniles commented on server/controllers/finance/reports/financial.employee.js:
```diff
@@ -71,8 +76,30 @@ async function build(req, res, next) {
+    // as the request is ordered by date, the last line item will
+    // also be the employee's balance for the search period
+    if (options.extractEmployee) {
+      if (creditorOperations.transactions.length) {
```
You can clean this code up a bit by using lodash's `_.last()`.
```js
const lastTxn = _.last(creditorOperations.transactions);
data.lastCum = lastTxn || { cumsum : 0 };
data.extratCreditorText = data.lastCum.cumsum >= 0
          ? 'FORM.LABELS.CREDIT_BALANCE' : 'FORM.LABELS.DEBIT_BALANCE';
```

jniles reviewed (commented):
@lomamech, the code looks pretty good.  I've just made a few comments on naming things.

I'm a bit concerned about opening balances - It seems like, if we filter by date, we should remember the debtor's opening balance at the start of the filtering.  Right now, it seems like it is always 0.  For example, if we have the following transactions:


Txn ID | Debit | Credit | Date | cumsum
-- | --:| --:| -- | --:
HBB123 | 0 | 50 | 1-Jan-2019 | -50
HBB150 | 25 | 0 | 2-Jan-2019 | -25
HBB231 | 0 | 30 | 3-Jan-2019 | -55
HBB232 | 0 | 80 | 4-Jan-2019 | -135
HBB301 | 15 | 0 | 5-Jan-2019 | -120

If we choose the dates from 3-Jan-2019 to 30-Jan-2019, it looks like this code will produce:

Txn ID | Debit | Credit | Date | cumsum
-- | --:| --:| -- | --:
HBB231 | 0 | 30 | 3-Jan-2019 | -30
HBB232 | 0 | 80 | 4-Jan-2019 | -110
HBB301 | 15 | 0 | 5-Jan-2019 | -95

When we should actually have:

Txn ID | Debit | Credit | Date | cumsum
-- | --:| --:| -- | --:
**Balance at 3-Jan-2019** | **25** | **50** | - | **-25**
HBB231 | 0 | 30 | 3-Jan-2019 | -55
HBB232 | 0 | 80 | 4-Jan-2019 | -135
HBB301 | 15 | 0 | 5-Jan-2019 | -120

Can you confirm that this is the expected behavior?  It seems important to me that we carry forward the balance, even if we filter out the transactions.

lomamech committed:
improvement(Employee Standing)
- Fix and Upgrate the translation key
- Adding opening balance until a date from
- Adding new function openingBlanceCreditor with parameters like
  creditorUuid and datefrom

lomamech committed:
Sanitaze code

jniles commented on server/controllers/finance/creditors.js:
```diff
@@ -126,22 +167,23 @@ function getFinancialActivity(creditorUuid) {
-    balance(creditorUuid),
-  ])
-    .spread((transactions, aggs) => {
+  const tabSql = (dateFrom && dateTo)
```
I think this representation might be cleaner:
```js
const tabSQL = [ db.exec(sql, [uid, uid]), balance(creditorUuid) ];
if (dateFrom && dateTo) { tabSQL.push(openingBalanceCreditor(creditorUuid, dateFrom)); }
```

jniles commented on server/controllers/finance/reports/financial.employee.handlebars:
```diff
@@ -43,6 +43,21 @@
+          <tr>
+            <td colspan="4" class="text-right"><b>{{translate "TABLE.COLUMNS.BALANCE_AT" }} {{date dates.dateFrom }} </b></td>
+            <td class="text-right">
+              {{debcred creditoropeningBalance.debit metadata.enterprise.currency_id}}
```
Can you capitalize each word here?  `creditorOpeningBalance` is better than `creditoropeningbalance`.

jniles reviewed (commented):
@lomamech this looks pretty good to me. Just two small code clarity changes and we can merge.

lomamech committed:
improvement(Employee Standing)
- clean the representation of tabSQL
- Capitalize creditoropeningbalance

lomamech committed:
clean code
